0003-I-dapr-compose.md

```diff
@@ -0,0 +1,294 @@
+  - `command` ["exec" "arg1" "arg2"] format for application command
+  - `env` which overrides or adds to common env var defined or the shell env var passed in when `dapr compose` is called
+  - `restart` field which lets `dapr compose` know if the app/sidecar should be restarted on crash or not
+
+The DAPR_HTTP_PORT and DAPR_GRPC_PORT will be passed in as extra environment variables to the application that is being run. Those flags for `daprd` can be overridden in the config file above but that is optional as the `compose` command will randomly assign a free port for each of those ports, similar to how `run` command currently behaves.
+
+### Preedence rules
+
+For `env` field:
+> Note: In addition to the defined env fields the app also gets the `DAPR_HTTP_PORT` and `DAPR_GRPC_PORT` fields.
+
+- If no field is present, the environment variables of the current shell which executes the `compose` command is passed to the app and dir.
+- if `env` field is present in the `common` section, in addition to the shell environment variables, the `env` map defined will be passed to all `apps` and `daprd` sidecars.
+- if `env` field is present only in a particular `apps` section, any shell environment variables, `env` maps from `common` section and the `env` map for the current app will be passed to both the `app` and `daprd`.
+- The more specific `env` key-value pairs will override the lesser specific ones i.e. `apps` section specific `env` key-value pairs will override the key-value pairs from the `common` section which will override the passed in shell environment variables.
+
+
+For each app in the `apps` section, the `resources_dir` and `config_file` values will be resolved in the following order:
+- If `reosurce_dir` and/or `config_file` fields are given for any `apps[i]` configuration use that value as is.
+- If not, check for `.dapr` folder in the **`apps[i].app_dir`** folder. If found, use the `.dapr/resources` folder for configuring the resources and `.dapr/config.yaml` for the `daprd` configuration file(argument `--config-file` in `daprd`).
+- If not, check if a `resources_dir` and/or `config_file` fields are defined in the `common` section of the compose configuration file. If so use those values for those fields.
+- If not, default to `~/.dapr/resources/` for `resources_dir` and `~/.dapr/config.yaml` for `config_file` values.
+
+
+## Proposed command format
+
+Given the configuration file defined above, there should be a way to use the configuration file and run the differnt commands and `daprd` with the configuration given in the file.
+
+For this there will be a `compose` command that will be defined in the `dapr` CLI.
+
+### Constraints
+
+`compose` should mimic what `run` does for the most part. The defaults should not be different from `run` command with any new configuration options having sane documented least permissions based defaults.
```

Comment:
Assuming default values will be provided for attributes like DAPR_HTTP_PORT, what is the mechanism for the user to identify this port, so that they could execute a HTTP request etc.? It would be good to document user experience for interacting with daprd launched with compose

Reply:
We should be mounting these env vars on the user process.

Reply:
Both DAPR_HTTP_PORT and DAPR_GRPC_PORT will be passed as env vars. I do not think metrics is port is needed inside the app, that can be part of the `dapr compose` output?